src/Raven.Server/Documents/ShardedTcpHandlers/ShardedSubscriptionConnection.cs

```diff
@@ -0,0 +1,394 @@
+            }
+
+            return result;
+        }
+
+        private bool CanStopSubscription(IReadOnlyDictionary<string, Exception> exceptions)
+        {
+            bool stopping = true;
+            foreach (var worker in _shardWorkers)
+            {
+                var ex = exceptions[worker.Key];
+                Debug.Assert(ex != null, "ex != null");
+
+                (bool shouldTryToReconnect, _) = worker.Value.Worker.CheckIfShouldReconnectWorker(ex, CancellationTokenSource, null, null, throwOnRedirectNodeNotFound: false);
+                if (shouldTryToReconnect)
+                {
+                    // we have at least one worker to try to reconnect
+                    stopping = false;
+                }
+            }
+
+            return stopping;
+        }
+
+        private async Task RedirectBatchAndConfirm(ShardedSubscriptionWorker.PublishedBatch batch, string shard)
+        {
+            try
+            {
+                // Send to client
+                await WriteBatchToClientAndAck(batch, shard);
+
+                // let sharded subscription worker know that we sent the batch to the client and received an ack request from it
+                batch.SendBatchToClientTcs.SetResult();
+            }
+            catch (Exception e)
+            {
+                // need to fail the shard subscription worker
+                batch.SendBatchToClientTcs.TrySetException(e);
+                throw;
+            }
+
+            // wait for sharded subscription worker to send ack to the shard subscription connection
+            // and receive the confirm from the shard subscription connection
+            await batch.ConfirmFromShardSubscriptionConnectionTcs.Task;
+
+            // send confirm to client and continue processing
+            await SendConfirmToClient();
+        }
+
+        private bool CanContinueSubscription(SubscriptionShardHolder shardHolder)
+        {
+            if (shardHolder.LastErrorDateTime.HasValue == false)
+            {
+                shardHolder.LastErrorDateTime = DateTime.UtcNow;
+                return true;
+            }
+
+            if (DateTime.UtcNow - shardHolder.LastErrorDateTime.Value <= _options.ShardedConnectionMaxErroneousPeriod)
+                return true;
+
+            return false;
+        }
+
+        private async Task WriteBatchToClientAndAck(ShardedSubscriptionWorker.PublishedBatch batch, string shard)
+        {
+            var replyFromClientTask = GetReplyFromClientAsync();
+            AddToStatusDescription(CreateStatusMessage(ConnectionStatus.Info, "Starting to send documents."));
+            int docsToFlush = 0;
+            string lastReceivedChangeVector = null;
+
+            var sendingCurrentBatchStopwatch = Stopwatch.StartNew();
+            using (TcpConnection.ContextPool.AllocateOperationContext(out JsonOperationContext context))
+            using (var buffer = new MemoryStream())
```

Comment:
I guess we can, but I used memory stream because its used in original code (_buffer is memory stream):
https://github.com/ravendb/ravendb/blob/f710aa237409fd0dad15fddd8c3b91a775edaa9d/src/Raven.Server/Documents/TcpHandlers/SubscriptionConnection.cs#L871

so I can reuse the same methods as we use in `SubscriptionConnection` where we 1st write in memory, then flush it to network stream:
https://github.com/ravendb/ravendb/blob/f710aa237409fd0dad15fddd8c3b91a775edaa9d/src/Raven.Server/Documents/TcpHandlers/SubscriptionConnection.cs#L327

Reply:
Can we reuse memory stream from context.CheckoutMemoryStream ?

Reply:
done